### PR TITLE
fix: NPCs hold fire if their target is out of max range

### PR DESCRIPTION
**Pull request by chaosvolt**
### Required

- [X] I wrote the PR title in [conventional commit format](https://docs.cataclysmbn.org/en/contribute/changelog_guidelines/).
- [X] I ran the [code formatter](https://docs.cataclysmbn.org/en/contribute/contributing/#code-style).
- [X] I linked any relevant issues using [github keyword syntax](https://docs.cataclysmbn.org/en/contribute/contributing/#pull-request-notes) so it can be closed automatically.
- [X] I have [committed my changes to new branch that isn't `main`](https://docs.cataclysmbn.org/en/contribute/contributing/#make-your-changes) so it won't cause conflict when updating `main` branch later.

## Purpose of change

This fixes cases where NPCs will waste shots on targets their gun literally lacks the range to hit.

## Describe the solution

Set `npc::method_of_attack` so that target distance being within their gun's max range is required before they commit to firing, so they won't repeatedly waste shots on targets they've dialed in on that their gun lacks the oomph to actually touch. Set here instead of `npc::confident_gun_mode_range` like I first considered doing, this way NPCs can still dial in and aim at targets outside their max range if they have the stats to try.

## Describe alternatives you've considered

screm

## Testing

1. Compiled and load-tested.
2. Handed NPC a flintlock pistol loaded with paper shot cartridges, maxed their shooting stats.
3. Spawned in a zombie a decent distance away and turned on debug mode.
4. NPC no longer fires prematurely.

### CODE REVIEW

**Review by scarf005 — COMMENTED** [review]
shouldn't `npc::confident_gun_mode_range` take account of gun's max range?

**Comment by chaosvolt** [comment]
> shouldn't `npc::confident_gun_mode_range` take account of gun's max range?

Main reason I didn't do that is because the aim action also checks for if the target's within confident gun range. Idea was that if confident range was higher than the gun's max range, and they acquire a target they can't hit yet, they'd then go ahead and try to improve their aim for when it does come into range (as the main issue that normally stops them from aiming and defaulting to "durr can't decide " is aim is already maxed well before they feel comfortable firing).

That said, the code also has it so that an NPC gets completely free aiming when pausing so this might not be essential? I'd keep it like this 100% if I was also stripping out the free aimbotting they get, which I do want to do at some point in the future too, but can move it to that function instead for now if desired.

**Comment by scarf005** [comment]
> Idea was that if confident range was higher than the gun's max range, and they acquire a target they can't hit yet, they'd then go ahead and try to improve their aim for when it does come into range (as the main issue that normally stops them from aiming and defaulting to "durr can't decide " is aim is already maxed well before they feel comfortable firing).

https://github.com/cataclysmbnteam/Cataclysm-BN/blob/e95f6098e05bb61219af53d3b9179fe8716e18b2/src/npcmove.cpp#L2024

if this is the intended effect, might be better to document it in function.

> That said, the code also has it so that an NPC gets completely free aiming when pausing so this might not be essential? I'd keep it like this 100% if I was also stripping out the free aimbotting they get, which I do want to do at some point in the future too, but can move it to that function instead for now if desired.

not sure if i understand this sentence correctly. could you provide some examples?

File: src/npcmove.cpp [hunk 1]
```diff
@@ -1380,7 +1380,7 @@ npc_action npc::method_of_attack()
 
     // if the best mode is within the confident range try for a shot
     if( g_mode && sees( *critter ) && has_los &&
-        confident_gun_mode_range( g_mode, cur_recoil ) >= dist ) {
+        confident_gun_mode_range( g_mode, cur_recoil ) >= dist && g_mode->gun_range( true ) >= dist ) {
```
Comment (scarf005) [inline src/npcmove.cpp:1383]:
```suggestion
        g_mode->gun_range( true ) >= dist && confident_gun_mode_range( g_mode, cur_recoil ) >= dist ) {
```

iirc `confident_gun_mode_range` is much more expensive to calculate than getting gun range so short-circuiting calculation could be better.

**Review by scarf005 — CHANGES_REQUESTED** [review]

**Comment by chaosvolt** [comment]
> not sure if i understand this sentence correctly. could you provide some examples?

This: https://github.com/cataclysmbnteam/Cataclysm-BN/blob/main/src/npcmove.cpp#L2619

```C++
    // NPCs currently always aim when using a gun, even with no target
    // This simulates them aiming at stuff just at the edge of their range
    if( !primary_weapon().is_gun() && cbm_active.is_null() ) {
        character_funcs::do_pause( *this );
        return;
    }

    // Stop, drop, and roll
    if( has_effect( effect_onfire ) ) {
        character_funcs::do_pause( *this );
    } else {
        aim();
        moves = std::min( moves, 0 );
    }
```

**Review by scarf005 — APPROVED** [review]

**Pull request merged**